Migrate RenderLayer to WeakKey hash tables

geoffreygaren opened the pull request:
#### 5ef0287a820b4ff1f131496361719b29759a9e89
<pre>
Migrate RenderLayer to WeakKey hash tables
<a href="https://bugs.webkit.org/show_bug.cgi?id=304034">https://bugs.webkit.org/show_bug.cgi?id=304034</a>
<a href="https://rdar.apple.com/166334471">rdar://166334471</a>

Reviewed by Sam Weinig.

This decouples RenderLayer from any particular WeakPtr implementation, so it can
switch to a more efficient implementation.

* Source/WTF/wtf/Forward.h: Declare canonical names for our new WeakKey hash
table types.

* Source/WTF/wtf/GetPtr.h:
* Source/WTF/wtf/HashMap.h:
(WTF::M&gt;::computeSize const):
(WTF::M&gt;::isEmptyIgnoringNullReferences const):
(WTF::M&gt;::removeWeakNullEntries):
* Source/WTF/wtf/HashSet.h:
(WTF::shouldValidateKey&gt;::computeSize const):
(WTF::shouldValidateKey&gt;::isEmptyIgnoringNullReferences const):
(WTF::shouldValidateKey&gt;::removeWeakNullEntries):
* Source/WTF/wtf/HashTable.h: Pushed weak table helper functions into HashTable
to create a single choke point where we decide to shrink a weak table based on
weak null keys.

(WTF::Malloc&gt;::computeSize const):
(WTF::Malloc&gt;::isEmptyIgnoringNullReferences const):
(WTF::Malloc&gt;::removeWeakNullEntries const):
(WTF::Malloc&gt;::removeIf):
(WTF::Malloc&gt;::takeIf):
(WTF::Malloc&gt;::computeBestTableSize):
(WTF::Malloc&gt;::checkTableConsistencyExceptSize const):
(WTF::Malloc&gt;::removeWeakNullEntries): Deleted.
* Source/WTF/wtf/ListHashSet.h: Boilerplate changes to make interfaces agnostic
about whether you look up by RenderLayer* or RenderLayer&amp;. This reduces
RenderLayer code churn in this patch.

Canonical link: <a href="https://commits.webkit.org/304372@main">https://commits.webkit.org/304372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/008cd93157cd873121e59255302f984c74ceb253

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135387 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7766 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46664 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143078 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d9edab6c-4b2d-44b6-8030-1c5a56ac2697) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137256 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8402 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7613 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a8a325ae-3e1b-4563-ac0f-9931708f1857) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6024 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121351 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84338 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3411 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3690 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127382 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/115008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39536 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145834 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133871 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7449 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40107 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111841 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7490 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6247 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112211 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5658 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117652 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61359 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20878 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7503 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35776 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166713 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7251 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71052 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7471 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7353 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -183,14 +183,18 @@ class HashMap final {
     MappedTakeType takeFirst();
 
+    // Useful when the key type is WeakPtr
+    template<typename = void> requires (KeyTraits::hasIsWeakNullValueFunction) size_t computeSize() const;
```
Presumably it is necessary (otherwise you wouldn't have added it), but that does the `template<typename = void>` do here? Is it needed for the requires statement to work?

geoffreygaren replied:
Yeah, it's weird. My understanding is that `requires` constrains template substitution, so there is no way to have a naked `requires` without an associated template. So `template<typename = void>` is the hack that means "I had to put some template here, so I'll use a no-op template".

Definitely open to ideas to make this less weird!

cdumez replied:
Following up in https://github.com/WebKit/WebKit/pull/55380.

weinig commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -213,36 +217,22 @@ class HashMap final {
+    template<SmartPtr K = KeyType> MappedTakeType take(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>*);
+
+    // Overloads for smart pointer keys that take the raw reference type as the parameter.
+    template<SmartPtr K = KeyType> iterator find(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>& ref) LIFETIME_BOUND { return find(&ref); }
```
What is the reason behind removing the use of the NonNullableSmartPtr concept?

geoffreygaren replied:
Previously, you could use `T*` APIs with either nullable or non-nullable smart pointers, but you could only use `T&` API with non-nullable smart pointers. `RenderLayer` + `WeakPtr` = code that likes to use `T&` + nullable smart pointers. 

I couldn't discern any rationale for requiring a non-nullable smart pointer in order to use a `T&` API, so I opened the APIs to all smart pointers, and that seems to work.

geoffreygaren replied:
I originally wrote one copy of the `T&` APIs for `NullableSmartPtr` and a separate copy for `NonNullableSmartPtr`, but then I noticed they were identical and collapsed them together.

weinig commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1560,6 +1561,8 @@ struct SizedUnsignedTrait<8> {
 template<typename T>
 using SameSizeUnsignedInteger = SizedUnsignedTrait<sizeof(T)>::Type;
 
+static auto dereference = std::views::transform([](auto&& x) -> decltype(auto) { return *x; });
```
I would love for this to have a name that indicates it is a ranges view. Maybe `dereference_view`? A comment explaining it would also be good, since this kind of thing is still pretty uncommon in the codebase (I was taken aback by it being a `static auto` declaration, and had to lookup what std::views::transform did to really understand it).

weinig replied:
Maybe putting it in a WTF::views namespace?

weinig replied:
Its uses are really cool!

geoffreygaren replied:
I guess `dereferenceView` for WebKit style.

weinig commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2927,7 +2927,7 @@ void LocalFrameView::repaintSlowRepaintObjects()
     // Renderers with fixed backgrounds may be in compositing layers, so we need to explicitly
     // repaint them after scrolling.
-    for (auto& renderer : *m_slowRepaintObjects)
+    for (auto& renderer : *m_slowRepaintObjects | dereference)
```
A problem I have with reading this code is that it's not clear to me why `dereference` is ok. I usually think of dereferencing as something that I should probably null check before doing. In this case, obviously, that is not an issue, since the set will weed those out, but, it is still more confusing than one would hope for I think. Perhaps there is a way to make this more clear in the future.

geoffreygaren replied:
Yeah, I think it's worth some thought.

I considered having the iterator itself do the dereference by adding a new `IteratorKeyPeekType` to `HashTraits`. I worried that doing that would make it impossible for clients who cared to access the underlying WeakPtr key. But maybe that's fine? Like, you're not allowed to modify the key anyway. And in any context where you'd want to check the null-ity of the key a second time after a side-effecting operation, you'd probably have to make a local weak pointer copy anyway. And there's not much else you can do with a WeakPtr.

So maybe `IteratorKeyPeekType` is the way to go in the future. (`IteratorPeekType` would be wrong, though. You want to be able to modify a weak pointer when it's a value in a hash table.)

geoffreygaren committed:
Migrate RenderLayer to WeakKey hash tables
https://bugs.webkit.org/show_bug.cgi?id=304034
rdar://166334471

Reviewed by Sam Weinig.

This decouples RenderLayer from any particular WeakPtr implementation, so it can
switch to a more efficient implementation.

* Source/WTF/wtf/Forward.h: Declare canonical names for our new WeakKey hash
table types.

* Source/WTF/wtf/GetPtr.h:
* Source/WTF/wtf/HashMap.h:
(WTF::M>::computeSize const):
(WTF::M>::isEmptyIgnoringNullReferences const):
(WTF::M>::removeWeakNullEntries):
* Source/WTF/wtf/HashSet.h:
(WTF::shouldValidateKey>::computeSize const):
(WTF::shouldValidateKey>::isEmptyIgnoringNullReferences const):
(WTF::shouldValidateKey>::removeWeakNullEntries):
* Source/WTF/wtf/HashTable.h: Pushed weak table helper functions into HashTable
to create a single choke point where we decide to shrink a weak table based on
weak null keys.

(WTF::Malloc>::computeSize const):
(WTF::Malloc>::isEmptyIgnoringNullReferences const):
(WTF::Malloc>::removeWeakNullEntries const):
(WTF::Malloc>::removeIf):
(WTF::Malloc>::takeIf):
(WTF::Malloc>::computeBestTableSize):
(WTF::Malloc>::checkTableConsistencyExceptSize const):
(WTF::Malloc>::removeWeakNullEntries): Deleted.
* Source/WTF/wtf/ListHashSet.h: Boilerplate changes to make interfaces agnostic
about whether you look up by RenderLayer* or RenderLayer&. This reduces
RenderLayer code churn in this patch.

Canonical link: https://commits.webkit.org/304372@main
